bach/bach/_testing/utils.py

```diff
@@ -0,0 +1,116 @@
+import uuid
+from copy import copy
+from decimal import Decimal
+from typing import List, Union, Any
+
+import pandas as pd
+from sqlalchemy.engine import Engine
+
+from bach import DataFrame, Series
+from sql_models.util import is_postgres, is_bigquery, is_athena
+
+
+def _get_view_sql_data(df: DataFrame):
+    sql = df.view_sql()
+    # escape sql, as conn.execute will think that '%' indicates a parameter
+    sql = sql.replace('%', '%%')
+    with df.engine.connect() as conn:
+        db_rows = conn.execute(sql)
+
+    column_names = list(db_rows.keys())
+    db_values = [list(row) for row in db_rows]
+    print(db_values)
+    return column_names, db_values
+
+
+def _get_to_pandas_data(df: DataFrame):
+    pdf = df.to_pandas()
+    # Convert pdf to the same format as _get_view_sql_data gives
+    column_names = (list(pdf.index.names) if df.index else []) + list(pdf.columns)
+    pdf = pdf.reset_index()
+    db_values = []
+    for value_row in pdf.to_numpy().tolist():
+        db_values.append(value_row if df.index else value_row[1:])  # don't include default index value
+    print(db_values)
+    return column_names, db_values
+
+
+def _convert_uuid_expected_data(engine: Engine, data: List[List[Any]]) -> List[List[Any]]:
+    """
+    Convert any UUID objects in data to string, if we represent uuids with strings in the engine's dialect.
+    """
+    if is_postgres(engine):
+        return data
+    if is_athena(engine) or is_bigquery(engine):
+        result = [
+            [str(cell) if isinstance(cell, uuid.UUID) else cell for cell in row]
+            for row in data
+        ]
+        return result
+    raise Exception(f'engine not supported {engine}')
+
+
+def assert_equals_data(
+    bt: Union[DataFrame, Series],
+    expected_columns: List[str],
+    expected_data: List[list],
+    order_by: Union[str, List[str]] = None,
+    use_to_pandas: bool = False,
+    round_decimals: bool = False,
+    decimal=4,
+    convert_uuid: bool = False,
+) -> List[List[Any]]:
+    """
+    Execute the sql of ButTuhDataFrame/Series's view_sql(), with the given order_by, and make sure the
```

Comment:
```suggestion
    Execute the sql of DataFrame/Series's view_sql(), with the given order_by, and make sure the
```

very old docstring :monocle_face:

Reply:
Fixed here https://github.com/objectiv/objectiv-analytics/pull/1250/commits/3274661888e5acc4b6c27288f6a871bb0ad0f538